File: atosim/org.palladiosimulator.blockchainsystems.threesim/src/main/java/org/palladiosimulator/blockchainsystems/threesim/metrics/calculators/DoubleSpendSuccessProbabilityCalculator.java
package org.palladiosimulator.blockchainsystems.threesim.metrics.calculators;

import org.palladiosimulator.blockchainsystems.threesim.metrics.DoubleSpendSuccessProbability;
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetric;
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetricImpl;
import java.util.List;

public class DoubleSpendSuccessProbabilityCalculator {
    private final double attackerHashPower;
    private final int confirmationDepth;

    public DoubleSpendSuccessProbabilityCalculator(double attackerHashPower, int confirmationDepth) {
        this.attackerHashPower = attackerHashPower;
        this.confirmationDepth = confirmationDepth;
    }

    public DoubleSpendSuccessProbability calculate() {
        double q = attackerHashPower;
        double p = 1 - q;
        if (q >= p) return new DoubleSpendSuccessProbability(1.0);
        double lambda = confirmationDepth * (q / p);
        double sum = 0;
        for (int k = 0; k <= confirmationDepth; k++) {
            double poisson = Math.exp(-lambda) * Math.pow(lambda, k) / factorial(k);
            double term = 1 - Math.pow(q / p, confirmationDepth - k);
            sum += poisson * term;
        }
        return new DoubleSpendSuccessProbability(1 - sum);
    }

    private double factorial(int n) {
        double r = 1.0;
        for (int i = 2; i <= n; i++) r *= i;
        return r;
    }

    public static AverageOutputMetric calculateAverage(List<DoubleSpendSuccessProbability> values) {
        double avg = values.stream().mapToDouble(DoubleSpendSuccessProbability::getValue).average().orElse(0.0);
        double variance = values.stream().mapToDouble(v -> (v.getValue() - avg) * (v.getValue() - avg)).average().orElse(0.0);
        double std = Math.sqrt(variance);
        Double cv = avg == 0.0 ? null : std / avg;
        return new AverageOutputMetricImpl(DoubleSpendSuccessProbability.NAME, avg, DoubleSpendSuccessProbability.UNIT, std, cv);
    }
}

File: atosim/org.palladiosimulator.blockchainsystems.threesim/src/main/kotlin/org/palladiosimulator/blockchainsystems/threesim/metrics/calculators/DoubleSpendSuccessProbabilityCalculator.kt
package org.palladiosimulator.blockchainsystems.threesim.metrics.calculators

import kotlin.math.exp
import kotlin.math.pow
import kotlin.math.sqrt
import org.palladiosimulator.blockchainsystems.threesim.metrics.DoubleSpendSuccessProbability
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetric
import org.palladiosimulator.blockchainsystems.threesim.metrics.abstractions.AverageOutputMetricImpl

class DoubleSpendSuccessProbabilityCalculator(
    private val attackerHashPower: Double,
    private val confirmationDepth: Int
) {

    fun calculate(): DoubleSpendSuccessProbability {

        val q = attackerHashPower
        val p = 1 - q

        if (q >= p) {
            return DoubleSpendSuccessProbability(1.0)
        }

        val lambda = confirmationDepth * (q / p)

        var sum = 0.0

        for (k in 0..confirmationDepth) {

            val poisson = poisson(lambda, k)

            val term = 1 - (q / p).pow((confirmationDepth - k).toDouble())

            sum += poisson * term
        }

        val probability = 1 - sum

        return DoubleSpendSuccessProbability(probability)
    }

    private fun poisson(lambda: Double, k: Int): Double {
        return exp(-lambda) * lambda.pow(k.toDouble()) / factorial(k)
    }

    private fun factorial(n: Int): Double {
        var result = 1.0
        for (i in 2..n) result *= i
        return result
    }

    companion object {

        fun calculateAverage(values: List<DoubleSpendSuccessProbability>): AverageOutputMetric {

            val avg = values.map { it.value }.average()

            val variance = values
                .map { (it.value - avg) * (it.value - avg) }
                .average()

            val std = sqrt(variance)

            val cv = if (avg == 0.0) null else std / avg

            return AverageOutputMetricImpl(
                name = DoubleSpendSuccessProbability.NAME,
                average = avg,
                unit = DoubleSpendSuccessProbability.UNIT,
                standardDeviation = std,
                coefficientOfVariation = cv
            )
        }
    }
}